### PR TITLE
Fixes with comments from the community

### PR DESCRIPTION
**Pull request by rubenrua**
New commit with the comments of @vlakoff and @Pierstoval in pull request #6168.

### CODE REVIEW

**Comment by naderman** [comment]
I don't think this makes any practical difference, isset and is_array are not expensive. Please provide benchmarks for performance optimizations.

**Comment by vlakoff** [comment]
Indeed, my code comment was just a thought. It would need proper benchmarking, and probably would be noticeable only with *many* method calls.

File: src/Composer/DependencyResolver/RuleSet.php [hunk 1]
```diff
@@ -60,17 +60,19 @@ public function add(Rule $rule, $type)
...
 
+        $rulesByHash = isset($this->rulesByHash[$hash]) ? $this->rulesByHash[$hash] : null;
+        $isArray = $rulesByHash && is_array($rulesByHash);
+
         // Do not add if rule already exists
-        if (isset($this->rulesByHash[$hash])) {
-            $potentialDuplicates = $this->rulesByHash[$hash];
-            if (is_array($potentialDuplicates)) {
-                foreach ($potentialDuplicates as $potentialDuplicate) {
+        if ($rulesByHash) {
```
Comment (alcohol) [inline src/Composer/DependencyResolver/RuleSet.php]: Why not `null !== $rulesbyHash`? No need for type coercion then.

File: src/Composer/DependencyResolver/RuleSet.php [hunk 2]
```diff
@@ -86,12 +88,12 @@ public function add(Rule $rule, $type)
 
         $this->nextRuleId++;
 
-        if (!isset($this->rulesByHash[$hash])) {
+        if (!$rulesByHash) {
```
Comment (alcohol) [inline src/Composer/DependencyResolver/RuleSet.php]: Same as previous.

File: src/Composer/DependencyResolver/RuleSet.php (hunk 1 above)
Comment (rubenrua) [inline src/Composer/DependencyResolver/RuleSet.php]: Done

File: src/Composer/DependencyResolver/RuleSet.php (hunk 2 above)
Comment (rubenrua) [inline src/Composer/DependencyResolver/RuleSet.php]: Done

**Comment by rubenrua** [comment]
@naderman The goal of this pull request is not performance optimization. IMHO the changes are good and they try to show the composer code style (very optimized code).

> You can optimise for execution speed. You can optimise for space. But the most precious thing you should optimise for is understandability

Feel free to close it. :)

**Comment by naderman** [comment]
That's really not the composer code style. Composer is optimized yes, in places where it is needed. But otherwise we try to make code as readable as possible, and this PR is not optimization like you said. There's nothing gained from any of these changes. (The triple equal is fine by itself, but has no practical effect either since both sides are guaranteed to be integers)

So thanks, but let's try and focus on actually improving the performance as in the original PR this was based on :-)

**Pull request closed without merging**